src/runtime/d3d12compute.cpp

```diff
@@ -0,0 +1,2794 @@
+class StackPrinter : public Printer<type, length> {
+public:
+    StackPrinter(void *ctx = NULL) : Printer<type, length>(ctx, buffer) { }
+    StackPrinter& operator () (void *ctx = NULL) {
+        this->user_context = ctx;
+        return *this;
+    }
+    uint64_t capacity() const { return length; }
+private:
+    char buffer [length];
+};
+
+// v trace and logging utilities for debugging v
+// ! definitely not super thread-safe stuff... !
+// in case there's no 'user_context' available in the scope of a function:
+static void *const user_context = NULL;
+//
+#if HALIDE_D3D12_TRACE
+    #ifndef DEBUG_RUNTIME
+    #define debug(x) Printer<BasicPrinter>(x)
+    #endif//DEBUG_RUNTIME
+    static volatile int indent_lock = 0;
+    static const char indent_pattern [] = "   ";
+    static char  indent [2048] = { };
+    static int   indent_end   = 0;
+    struct TraceLogScope {
+        TraceLogScope() {
+            while (__sync_lock_test_and_set(&indent_lock, 1)) {}
+            for (const char *p = indent_pattern; *p; ++p) {
+                indent[indent_end++] = *p;
+            }
+            __sync_lock_release(&indent_lock);
+        }
+        ~TraceLogScope() {
+            while (__sync_lock_test_and_set(&indent_lock, 1)) {}
+            for (const char *p = indent_pattern; *p; ++p) {
+                indent[--indent_end] = '\0';
+            }
+            __sync_lock_release(&indent_lock);
+        }
+    };
+    #define TRACEINDENT     ((const char*)indent)
+    #define TRACEPRINT(msg) debug(user_context) << TRACEINDENT << msg;
+    #define TRACELOG        TRACEPRINT("[@]" << __FUNCTION__ << "\n"); TraceLogScope trace_scope___;
+#else
+    #define TRACEINDENT ""
+    #define TRACEPRINT(msg)
+    #define TRACELOG
+#endif
+//
+#define ERRORLOG    error(user_context) << TRACEINDENT
+// ^ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ^
+
+static const char* d3d12_debug_dump();
+
+#define d3d12_halt(...)             \
+    ERRORLOG << __VA_ARGS__ << "\n" \
+             << d3d12_debug_dump()  \
+             << "!!! HALT !!!\n"
+
+void *d3d12_load_library(const char *name) {
+    TRACELOG;
+    void *lib = halide_load_library(name);
+    if (lib) {
+        TRACEPRINT("Loaded runtime library '" << name << "' at location " << lib << "\n");
+    } else {
+        d3d12_halt("Unable to load runtime library: " << name);
+    }
+    return lib;
+}
+
+void *d3d12_get_library_symbol(void *lib, const char *name) {
+    TRACELOG;
+    void* symbol = halide_get_library_symbol(lib, name);
+    if (symbol) {
+        TRACEPRINT("Symbol '" << name << "' found at " << symbol << "\n");
+    } else {
+        d3d12_halt("Symbol not found: " << name);
+    }
+    return symbol;
+}
+
+#define d3d12_debug_break (*((volatile int8_t*)NULL) = 0)
+
+#ifndef UNUSED
+#define UNUSED(x) ((void)x)
+#endif//UNUSED
+
+#if HALIDE_D3D12_RENDERDOC
+#if HALIDE_D3D12_DEBUG_LAYER
+    #pragma message "RenderDoc might now work well along with the Dirct3D debug layers..."
```

Comment:
"now work well" or "not work well"?

Reply:
Used to be "not" but recent versions of RenderDoc now support it; I'll rephrase it.